libfreerdp/crypto/ssl.c

```diff
@@ -0,0 +1,388 @@
+	}
+	else
+	{
+		CRYPTO_set_dynlock_create_callback(_freerdp_openssl_dynlock_create);
+		CRYPTO_set_dynlock_lock_callback(_freerdp_openssl_dynlock_lock);
+		CRYPTO_set_dynlock_destroy_callback(_freerdp_openssl_dynlock_destroy);
+	}
+
+	/* Use the deprecated CRYPTO_get_id_callback() if building against OpenSSL < 1.0.0 */
+#if (OPENSSL_VERSION_NUMBER < 0x10000000L)
+
+	if (CRYPTO_get_id_callback())
+	{
+		WLog_WARN(TAG, "OpenSSL id_callback is already set");
+	}
+	else
+	{
+		CRYPTO_set_id_callback(_freerdp_openssl_id);
+	}
+
+#endif
+	return TRUE;
+}
+
+static BOOL _freerdp_openssl_cleanup_locking(void)
+{
+	/* undo our static locking modifications */
+	if (CRYPTO_get_locking_callback() == _freerdp_openssl_locking)
+	{
+		int i;
+		CRYPTO_set_locking_callback(NULL);
+
+		for (i = 0; i < g_freerdp_openssl_num_locks; i++)
+		{
+			CloseHandle(g_freerdp_openssl_locks[i]);
+		}
+
+		g_freerdp_openssl_num_locks = 0;
+		free(g_freerdp_openssl_locks);
+		g_freerdp_openssl_locks = NULL;
+	}
+
+	/* unset our dynamic locking callbacks */
+
+	if (CRYPTO_get_dynlock_create_callback() == _freerdp_openssl_dynlock_create)
+	{
+		CRYPTO_set_dynlock_create_callback(NULL);
+	}
+
+	if (CRYPTO_get_dynlock_lock_callback() == _freerdp_openssl_dynlock_lock)
+	{
+		CRYPTO_set_dynlock_lock_callback(NULL);
+	}
+
+	if (CRYPTO_get_dynlock_destroy_callback() == _freerdp_openssl_dynlock_destroy)
+	{
+		CRYPTO_set_dynlock_destroy_callback(NULL);
+	}
+
+#if (OPENSSL_VERSION_NUMBER < 0x10000000L)
+
+	if (CRYPTO_get_id_callback() == _freerdp_openssl_id)
+	{
+		CRYPTO_set_id_callback(NULL);
+	}
+
+#endif
+	return TRUE;
+}
+
+#endif /* OpenSSL < 1.1.0 */
+
+static BOOL CALLBACK _freerdp_openssl_initialize(PINIT_ONCE once, PVOID param, PVOID* context)
+{
+	DWORD flags = param ? *(PDWORD)param : WINPR_SSL_INIT_DEFAULT;
+
+	if (flags & WINPR_SSL_INIT_ALREADY_INITIALIZED)
+	{
+		return TRUE;
+	}
+
+#ifdef WINPR_OPENSSL_LOCKING_REQUIRED
+
+	if (flags & WINPR_SSL_INIT_ENABLE_LOCKING)
+	{
+		if (!_freerdp_openssl_initialize_locking())
+		{
+			return FALSE;
+		}
+	}
+
+#endif
+	/* SSL_load_error_strings() is void */
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
+	SSL_load_error_strings();
+	/* SSL_library_init() always returns "1" */
+	SSL_library_init();
+	OpenSSL_add_all_digests();
+	OpenSSL_add_all_ciphers();
+#else
+
+	if (OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS | OPENSSL_INIT_LOAD_CRYPTO_STRINGS |
+	                         OPENSSL_INIT_ADD_ALL_CIPHERS | OPENSSL_INIT_ADD_ALL_DIGESTS |
+	                         OPENSSL_INIT_ENGINE_ALL_BUILTIN,
+	                     NULL) != 1)
+		return FALSE;
+
+#endif
+	g_freerdp_openssl_initialized_by_winpr = TRUE;
+
+	if (flags & WINPR_SSL_INIT_ENABLE_FIPS)
+	{
+#if (OPENSSL_VERSION_NUMBER < 0x10001000L) || defined(LIBRESSL_VERSION_NUMBER)
+		WLog_ERR(TAG,
+		         "Openssl fips mode ENable not available on openssl versions less than 1.0.1!");
+#else
```

Comment:
Typo "ENable"

Reply:
I fixed those typos and made the debug logging casing more consistent, this was basically already there.